core/build/cli.odin

```diff
@@ -0,0 +1,182 @@
+package build
+
+import "core:strings"
+import "core:os"
+import "core:path/filepath"
+import "core:sync"
+import "core:encoding/json"
+import "core:runtime"
+import "core:fmt"
+import "core:slice"
+import "core:unicode/utf8"
+import "core:unicode"
+import "core:strconv"
+import "core:intrinsics"
+
+
+Flag_Arg :: struct { // -flag:key=val
+	flag: string,
+	key: string,
+	val: string,
+}
+
+Arg :: union {
+	Flag_Arg,
+	string,
+}
+
+Args_Error :: enum {
+	None,
+	Invalid_Format,
+}
+
+// Todo(Dragos): add option to handle ints, floats, bools as strings
+parse_args :: proc(os_args: []string, allocator := context.allocator) -> (args: []Arg, err: Args_Error) {
+	runtime.DEFAULT_TEMP_ALLOCATOR_TEMP_GUARD(ignore = allocator == context.temp_allocator)
+	context.allocator = allocator
+	args = make([]Arg, len(os_args) - 1)
+	for os_arg, i in os_args[1:] {
+		if os_arg[0] == '-' {
+			flag_arg: Flag_Arg
+			colon_slice := strings.split(os_arg, ":", context.temp_allocator)
+			switch len(colon_slice) {
+			case 1: // only the flag found, no key-val
+				flag_arg.flag = colon_slice[0]
+				
+			case 2: // key and/or value found
+				flag_arg.flag = colon_slice[0]
+				equal_slice := strings.split(colon_slice[1], "=", context.temp_allocator)
+				switch len(equal_slice) {
+				case 1: // only key, no value
+					flag_arg.key = equal_slice[0]
+				case 2: /// key and value
+					flag_arg.key = equal_slice[0]
+					flag_arg.val = equal_slice[1]
+				}
+			
+			case: // more than 1 colon found. Invalid syntax
+				err = .Invalid_Format
+				return
+			}
+			args[i] = flag_arg
+		} else {
+			args[i] = os_arg
+		}
+	}
+	return
+}
+
+
+Flag_Desc :: struct {
+	mode: Run_Mode,
+	flag: Flag_Arg,
+	help: string,
+}
+
+Cli_Info :: struct {
+	project: ^Project, // this can be projects: []^Project
+	default_target: ^Target,
+}
+
+Cli_Flags_Error :: enum {
+	Incompatible_Flags,
+}
+
+builtin_flags := [?]Flag_Desc {
+	{.Help, {"-help", "", ""}, "Displays information about the build system or the target specified."},
+	{.Dev, {"-ols", "", ""}, "Generates an ols.json for the configuration."},
+	{.Dev, {"-vscode", "", ""}, "Generates .vscode folder for debugging."},
```

Comment:
Technically `ols` is unofficial, so having part of an official tool is kind of "weird". Same with the `-vscode` thing. Yes both are very useful, but _I am not sure_ if those should be here or not.

Reply:
yeah i understand that. The flags you pass to your build system can be expanded by yourself, so the builtin flags like `-ols` and `-vscode` are things that the user would write themselves anyway, so why not provide them as "sensible defaults". Generally speaking, all `.Dev` flags are meant for "non-official" dev tools, like setting up your debugger, your language server, anything you need to make your development easier. Personally i think this approach is more useful than it's weird, and I'd like to keep expanding on the `.Dev` flags for other editors and tools that people use (and can be used in a "generic" way like configuring a json). The usefulness itself also comes from the way `core:build` in it's current version works, as you are able to configure your dev env specifically for a certain target

Reply:
so I've been thinking about this. The tools to generate 3rd party things is quite useful, but indeed there is a bit of "weirdness" into it. I do think the feature itself should stay, as it makes `core:build` more useful than a python/bat/sh script, so here are some things that I could experiment with if you end up not agreeing with 3rd party being in `core`:
- Make a `vendor:build/dev` package, or alternatively different packages for each 3rd party tool `vendor:build/vscode` `vendor:build/ols`. These can handle setting up default development environments that we support "officially" via the `vendor` packages. As `vendor` is already supposed to contain 3rd party things, this solution would make the most sense.
- Create a system to "register" dev environments, something like `build.add_dev("vscode", flags_that_trigger_vscode_gen, vscode_gen_proc)`. This could also work well with the addition of optional `vendor:build` already made dev environments.

My biggest gripe with any of these solutions is that ultimately I wanted `core:build` to be a single self contained package, in order to allow importing build systems into eachother, a feature which is very useful when you want to build multiple projects as a single one.

Reply:
while you think about it, I'll try to handle flags a bit better, and at least separate the third party things better, and we'll see what we do with them. Right now all the `.Dev` flags are bound to 3rd party generation, so probably the best way to separate them for now is to make the `.Dev` mode triggered only by a `-dev` flag